_posts/2022-05-23-fine-grained-access-control-for-multitenancy.markdown

```diff
@@ -0,0 +1,326 @@
+  apiId: 'htieqffa0l'
+}
+```
+
+#### Multi-tenant DynamoDB table
+
+In the context of dynamoDb, the tenant ID will essentially be the partition key that'll be used to group together/ partition customer data. When a customer needs to write/retrieve data to/from the purchase history database they'll use their unique tenant ID to only access data that belongs to them. In essence the tenant ID can be thought of as being analogous to a key that can only open a single door.
+
+The table provided below is a representation of how the data stored in the purchase history will be partitioned. The parameter `tenantId` is used as the `partition key` while the `itemId` parameter is used as the `sort key`.
+
+|     tenantID     | itemID      |  price   | qty | description |
+| :--------------: | :---------- | :------: | :-: | :---------: |
+| `Customer1-xcv9` | `timestamp` | `number` |     |  `string`   |
+| `Customer2-dgf1` | `timestamp` | `number` |     |  `string`   |
+| `Customer1-xcv9` | `timestamp` | `number` |     |  `string`   |
+| `Customer2-dgf1` | `timestamp` | `number` |     |  `string`   |
+
+### Fine Grained Access Control
+
+When the Lambda Authorizer is invoked it is expected to return a JSON object that not only includes a context object that contains the `tenantID` parameter but also a resource policy as detailed [here](#lambda-authorizer-output-sample). The policy should allow/deny access to the API and dynamoDB table depending on the outcome of the Lambda Authorizer's token claims verification.
+
+All Allow-Policies generated will allow the authenticated user to only access data that has a partition key which matches their tenant id. This is achieved by making use of DynamoDB's `dynamodb:LeadingKeys` condition key for fine grained access control.
+In essence, the `dynamodb:LeadingKeys` condition key is a mechanism for row level access control within DynamodDB tables. Employing the use of this mechanism guarantees a reliable and secure multitenant environment as well as the added benefit of a simplified data persistence layer.
+
+![image](dynamdbFineGrainedAccess.png)
+
+#### Lambda Authorizer Resource Policies
+
+###### Token Verification failed
+
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Deny",
+            "Action": "*",
+            "Resource": "*"
+        },
+    ]
+}
+```
+
+###### Token Verification passed
+
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "execute-api:Invoke",
+            "Resource": executeApiArn,
+        },
+        {
+
+          Effect: 'Allow',
+          Action: ['dynamodb:UpdateItem', 'dynamodb:PutItem', 'dynamodb:DeleteItem', 'dynamodb:Query'],
+          Resource: this.cartTable,
+          Condition: {
+            'ForAllValues:StringLike': {
+              'dynamodb:LeadingKeys': tenantId,
+            },
+          },
+        },
+    ]
+}
+```
```

Comment:
Suggestion; conventionally we finish posts off with a summary. Re-cap what you covered and achieved in the tutorial, and what next steps might be. Point them to our contact page if they need consulting work.